Add ReadonlyArray overload to Array.isArray

ulrichb opened the pull request:
Fixes #17002

This is an alternative to #22942 which doesn't introduce a breaking change for `any` inputs by using an overload instead of changing the `Array.isArray()` signature.

Therefore it still works for `any`/`unknown` arguments to cast them to `Array<any>` (instead of  `ReadonlyArray<any>` like in #22942). See the test cases in `isArray.ts`.


In the following case (the argument is some _non-array_ + _non-`any`/`unknown`_) this PR still introduces a breaking change.

```ts
const notReallyAnArray = {};

if (Array.isArray(notReallyAnArray)) {
    notReallyAnArray.length; // OK
    const str: string = notReallyAnArray[0]; // OK
    notReallyAnArray.sort(); // Now an Error
}
```

IMO it's okay to break this situation (where the static type isn't an array and should be changed anyways) and on the other hand we win type safe `ReadonlyArray<T>` narrowing (#17002, which by the way is a breaking change anyways).

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=28916) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=28916) before we can accept your contribution.<br/><br/>:x: ulrichb [sign now](https://cla.opensource.microsoft.com/Microsoft/TypeScript?pullRequest=28916)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/TypeScript?pullRequest=28916) it.</sub>

RyanCavanaugh commented:
@sandersn a lot to unpack here; we should discuss beforemerging

MayhemYDG commented:
@ulrichb @RyanCavanaugh 
Would love to see this fixed, currently `Array.isArray` breaks types in code like this:
```ts
function test(arg: string | ReadonlyArray<string>): string {
  if (Array.isArray(arg)) {
    return arg.join('\n');
  }
  // Type 'string | readonly string[]' is not assignable to type 'string'.
  //  Type 'readonly string[]' is not assignable to type 'string'.ts(2322)
  return arg;
}
```

It's fine with `arg: string | Array<string>` of course, but that's not ideal.

RyanCavanaugh commented:
interface MyArray<T> extends Array<T> { manifest: any; }
interface MyReadOnlyArray<T> extends ReadonlyArray<T> { manifest: any; }

namespace Testbench {
    declare const Array: {
        isArray<T>(arg: T): arg is T extends ReadonlyArray<any> ? T : Array<any> extends T ? (T & any[]) : never;
    };

    function fn1(arg: string | string[]) {
        if (Array.isArray(arg)) arg.push(""); // Should OK
    }

    function fn2(arg: unknown) {
        if (Array.isArray(arg)) arg.push(""); // Should OK
    }

    function fn3(arg: object) {
        if (Array.isArray(arg)) arg.push(""); // Should OK
    }

    function fn4(arg: {}) {
        if (Array.isArray(arg)) arg.push(""); // Should OK
    }

    function fn5(arg: string | ReadonlyArray<string>) {
        if (Array.isArray(arg)) arg.push(10); // Should FAIL
        if (Array.isArray(arg)) arg.push(""); // Should FAIL
    }

    function fn6(arg: string | string[]) {
        if (Array.isArray(arg)) arg.push(10); // Should FAIL
    }

    function fn7(arg: boolean | number[] | string[]) {
        if (Array.isArray(arg)) arg.push(null as any as string & number); // Should OK
    }

    function fn8(arg: string | number[] | readonly string[]) {
        if (Array.isArray(arg)) arg.push(10); // Should FAIL
    }

    function fn9(arg: string | number[] | readonly string[]) {
        if (Array.isArray(arg)) arg.push(10); // Should FAIL
    }

    function fn10(arg: string | MyArray<string>) {
        if (Array.isArray(arg)) arg.push(10); // Should FAIL
        if (Array.isArray(arg)) arg.push(""); // Should OK
        if (Array.isArray(arg)) arg.manifest; // Should OK
    }

    function fn11(arg: string | MyReadOnlyArray<string>) {
        if (Array.isArray(arg)) arg.push(""); // Should FAIL
        if (Array.isArray(arg)) arg.indexOf(10); // Should FAIL
        if (Array.isArray(arg)) arg.indexOf(""); // Should OK
        if (Array.isArray(arg)) arg.manifest; // Should OK
    }

    function fn12<T>(arg: T | T[]) {
        if (Array.isArray(arg)) arg.push(null as any as T); // Should OK
    }

    function fn13<T>(arg: T | ReadonlyArray<T>) {
        if (Array.isArray(arg)) arg.push(null as any as T); // Should fail
        if (Array.isArray(arg)) arg.indexOf(null as any as T); // Should fail
    }

    function fn14<T>(arg: T | [T]) {
        if (Array.isArray(arg)) arg.push(null as any as T); // Should OK
    }

    function fn15<T>(arg: T | readonly [T]) {
        if (Array.isArray(arg)) arg.push(null as any as T); // Should fail
        if (Array.isArray(arg)) arg.indexOf(null as any as T); // Should OK
    }

    function fn16<T extends string | string[]>(arg: T) {
        if (Array.isArray(arg)) arg.push("10"); // Should OK
    }

    function fn17() {
        const s: Array<string | string[]> = [];
        const arrs = s.filter(Array.isArray);
        arrs.push(["one"]);
    }
}

RyanCavanaugh commented:
@ulrichb we hammered out the definition above (see the top of the file); do you want to try updating the PR to use it and clean up the merge conflicts?

ulrichb commented:
Hi @RyanCavanaugh,

updated the PR and merged master.

I like the approach with the conditional type (avoids the breaking change mentioned in the description).

One question: Why not `isArray<T>(arg: T): arg is T extends ReadonlyArray<any> ? T : T & any[];` (no `Array<any> extends T` case and `never`)? I ran all tests locally and it behaves the equally.

A few small changes of the test cases I made:
* `fn5`: Added positive cases for ``ReadonlyArray` members
* `fn`:
* `fn`:
* `fn`:

ulrichb commented:
Hi @RyanCavanaugh,

updated the PR and merged master.

I like the approach with the conditional type (avoids the breaking change mentioned in the description).

One question: Why not `isArray<T>(arg: T): arg is T extends ReadonlyArray<any> ? T : T & any[];` (no `Array<any> extends T` case and `never`)? I ran all tests locally and it behaves equally.

----

A few small changes of the test cases I made:
* `fn5`: Added positive cases for `ReadonlyArray` members
* `fn13`: Fixed the wrong "should fail" comment
* `fn16/17`: Added additional negative test case
* Changed all the `as null as ...` casts to parameters.

ulrichb commented:
/ping @RyanCavanaugh

kalbert312 commented:
I was looking for a better type this as well. I'm using

```typescript
isArray<T>(arg: T): T is Extract<NonNullable<T>, Array | ReadonlyArray>;
```
I had to wrap `T` in `NonNullable` otherwise I was running into some issue with `undefined`.
Seems to work but only tested in one case.

sandersn commented:
Sorry, I lost track of this. I don't want to merge it during the 3.8 beta, but I added it to my list of work when the 3.9 release starts.

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+
+function fn13<T>(arg: T | ReadonlyArray<T>, t: T) {
+    if (Array.isArray(arg)) arg.push(t); // Should fail
+    if (Array.isArray(arg)) arg.indexOf(t); // OK
```
For consistency:
```suggestion
    if (Array.isArray(arg)) arg.indexOf(t); // Should OK
```

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+}
+
+function fn13<T>(arg: T | ReadonlyArray<T>, t: T) {
+    if (Array.isArray(arg)) arg.push(t); // Should fail
```
For consistency, let's make this (and all the other places):
```suggestion
    if (Array.isArray(arg)) arg.push(t); // Should FAIL
```

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
 
-if (Array.isArray(maybeArray)) {
-    maybeArray.length; // OK
+function fn1(arg: string | string[]) {
```
```suggestion
function fn0(arg: any) {
    if (Array.isArray(arg)) {
        arg.push(""); // Should OK
        arg.anything(); // Should FAIL
    } else {
        arg.anything(); // Should OK
    }
}

function fn1(arg: string | string[]) {
```

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+}
+
+function fn2(arg: unknown) {
+    if (Array.isArray(arg)) arg.push(""); // Should OK
```
```suggestion
    if (Array.isArray(arg)) {
        arg.push(""); // Should OK
        var n: number;
        n = arg[0]; // Should FAIL (or OK if you want `arg` to be narrowed to `any[]`)
    }
```

laughinghan commented on src/lib/es5.d.ts:
```diff
@@ -1341,7 +1341,7 @@ interface ArrayConstructor {
     <T>(arrayLength: number): T[];
     <T>(...items: T[]): T[];
-    isArray(arg: any): arg is Array<any>;
+    isArray<T>(arg: T): arg is T extends ReadonlyArray<any> ? T : Array<any> extends T ? (T & any[]) : never;
```
```suggestion
    isArray<T>(arg: T): arg is unknown extends T ? Extract<T[], T> : (T extends readonly any[] ? T : T & any[]);
```

laughinghan commented:
@RyanCavanaugh @ulrichb So there's an important and obvious test case that both of your versions fail:

```ts
  function fn0(arg: any) {
    if (isArrayTest(arg)) {
      arg.push(""); // Should OK
      arg.anything(); // Should FAIL, but both your versions pass,
                      // because `arg` is being typed as `any`
    } else {
      arg.anything(); // Should OK, but both your versions fail,
                      // because `arg` is being narrowed to `never`
    }
  }
```

Note that this is different from the `fn2()` test case where `arg` is `unknown`. Both of your typings (@ulrichb's is actually pretty easy to understand: `T extends readonly any[] ? T : T & any[]`) work for `unknown` because it is the proper, well-behaved top type, so it isn't assignable to `readonly any[]`, and so the type of `arg` gets correctly narrowed to `unknown & any[] = any[]`.

However, `any` is magic and acts like both the top and bottom types, so it **_is_** assignable to `readonly any[]`, and so the type of `arg` isn't narrowed at all and remains `any`, instead of being narrowed to `any[]`.

Of course, I didn't figure all that out until after much, much, much trial-and-error down endless dead-ends, but with all that in mind, my fix follows mostly straightforwardly:
```ts
unknown extends T ? Extract<T[], T> : (T extends readonly any[] ? T : T & any[])
```
Basically, I special-case `any` by checking if `unknown extends T`, which tells us that `T` must be either `any` or `unknown`, and if so, I ensure the type of `arg` gets narrowed to `any[]` or `unknown[]`, respectively.

Three subtleties:
- My type actually behaves slightly differently from both of yours in the `unknown` case (`fn2()`); yours both narrow the type of `arg` to `any[]`, whereas I think it should be narrowed to `unknown[]`. If you prefer your behavior, just change the `Extract<T[], T>` to `Extract<any[], T>`.
  To distinguish my behavior from yours, I extend the `fn2()` test case:
  ```ts
  function fn2(arg: unknown) {
    if (isArrayTest(arg)) {
      arg.push(""); // Should OK
      var n: number;
      n = arg[0]; // Should FAIL (or OK if you want `arg` to be narrowed to `any[]`)
    }
  }
  ```
- `Extract<T[], T>` must be used and not `T & T[]` because of my `fn0()` test case: if `T` is `any`, then `T & T[] = any & any[] = any & AnyTypeExpression = any`, whereas we want to narrow the type to `any[]`
  + (note that this is again because `any` is not a well-behaved top type but rather a mix of top and bottom, in this instance it's acting like the bottom type:  `never & AnyTypeExpression = never`, whereas `unknown & AnyTypeExpression = AnyTypeExpression`)
- `Extract<T[], T>` must be used and not `T[] extends T ? T : never`, but I don't understand why, [is this a TypeScript bug?](https://github.com/microsoft/TypeScript/issues/36135)

laughinghan commented:
OK so, my fix fails tests `fn12()` and `fn14()`, but I'd argue that those tests are wrong. They test that `T | T[]` is narrowed to `T[]` and `T | [T]` is narrowed to `[T]`, respectively, but that is actually incorrect if `T = readonly string[]`.

Note, however, that this is a potentially breaking change from the old `isArray(arg: any): arg is any[]` behavior, which did narrow `T | T[]` to `T[]` and `T | [T]` to `[T]`. Is that an intentionally unsoundness in TypeScript's type narrowing algorithm?

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+    if (Array.isArray(arg)) arg.push(10); // Should FAIL
+}
+
+function fn7(arg: boolean | number[] | string[], stringAndNumber: string & number) {
```
`string & number = never`, I don't think this tests what we want, how about:

```suggestion
function fn7(arg: boolean | Number[] | string[], stringAndNumber: String & Number) {
```

sandersn commented:
@laughinghan the problem you point about narrowing `any` is pretty severe. But I also don't think we want to break the narrowing testing in `fn12` and `fn14`, even if they are unsound.

laughinghan commented:
@sandersn I see. I will take a stab at it then

sandersn commented:
@laughinghan if you can come up with a solution that passes all the tests, that would help our discussion a lot. Otherwise that's what we'll spend our time doing first.

sandersn commented:
This PR hasn't seen any activity for quite a while, so I'm going to close it to keep the number of open PRs manageable. Feel free to open a fresh PR or continue the discussion here.

laughinghan commented:
@sandersn Oh—did it come up at the design meeting at all whether we want the `fn12` and `fn14` behavior, even though they're unsound

laughinghan commented:
@sandersn Actually, I fixed it! [See for yourself](https://www.typescriptlang.org/play/index.html?ssl=42&ssc=1&pln=43&pc=1#code/HYQwtgpgzgDiDGEAEAVaAXASgTxMAwiAG54gCuA5gBZQBqEATlAJYD2wSA3gFBJIBmZYPHRsOzKAEEGDENjRR0AHhQA+ABQgGFAFyoAlHq0UkE1EggAPdBGAATKEkwQQd9gBts02diV5sqkgA-OZ63nJ+wAEW1rYO5iHqKEgAZEj+ANoAuvpIesAQRIxcvHxIDBDoZAwc4dgAdBJ1mtr6ANylAL7cpYLCouwCwAAMLbrpUbk8Zab8SOpNMnIK6GP6U6UzxvUwZFBU6gBEh+1IAPRnSADKVKxk7nZIAPIA0ptl2-7oVMzAFOqnC7XW73R4AMUkAEkADLvToWdxQZDTGbpbT1L4-P4AtrnS43O4PZ5vGbdPjdbrcUCQWAIZArACq7gYzHgVAARnRGCxBii+iIxKYpEt5BgVBpjHoUIY0SYzMkrDZ7I5nK4PF4RZFoiFklLUhNsNkSjMKlUakg6o1hT41h1yb0hALBvwRmMjJNjWVmHMFtblhg1htUbKdnsDsdAfiQUTXu8+J8ot9fv9I8DCeCobDSQikZ6tujMcmcXi06DiXCutxKdToHBEKgMNDyNRk1Q8FAAAoMVgwVhQEDuT2-GwMfh0pAAWQ1PnFMSV8Tqs84SDAeG9GHd2FxZNMwBHY-rU9VdiewE8i7Uc7iKpcbjP04il+Xq+A68Um+3PT4-IG4j9osUcU3QMIxtCFJAAFFrFkEQtWyAAaVBAgAH0g6CEGUDJ-CyRDL1Q9QhAAa2AVgAHcOEVa8EiQAoigYPJ5gVWJlXKW91QNI0dQY2jGHWPMS2+ZB+GYJh0CQdBSNYJB4HccgkSgRCAAMoPQGDlEyHCkMUiZHmU9DYKwqJNLURT4PeIFbCgapkG+EAxME8TsBgZAYAqOxWTs5BSOYdxB0VNSkEUgAhbJtLuMTWDmRTJCQVCQqyRTzMuPBdIyIKEqQcLMqimLULShLEIAKz2MT3GYQjkBAcSqGQUBmCKJBXIgdz4E8wLjHAjTEpmIFJNBepUBqgRfgHaTZL2ZAzCq2AIHgZhRta3NvXMMxiLIjh0Ck34ZLIOwICS9JWLVe80R8RDNrnALFIqY7PCQeKwrICKctio67zuh6zJ65L7ECm73uwJB8se57Atyt72OBxD2Se9JHGm5y5tGqxEBgX8DuYMSCmaxwLuYCgSIqarkEW5BIvMABeA1odmuTkEU-xtKm8RfIAWnZCA2yKOwDvQJyIAGq4IGQKh0HQGAoB0C4KExqgyHZep4FYMAzjAVluygSL0DOFB+aueAWTRs5dl8s4ACYAA4AE4AEYADZ3lNapahFK1mmMdpKwdfpBRdUZJQNIMvR9RYfBWQN+PjdFdn2I4TlxIECTLWNgwTbAk2xVMk6JCEYThHNkTjENC0zhOo3TctSS979HV-IYbeAxQWT+V6m+TbIg74ZbfTqcOPdybYY-DeOS2zx4U6QSka5951gDN4CiJI8jO9meZQ-9RQI5RfMKFDWOIzL0sYxJVESHo4B8jIMAOYYO1UQ4KnjAyYYskPsfqrMV5Mvo3PoSCIvgzBiBJCDgrBBL0UkgwBwiF9gVw6lzZAHNbAAMAd9GiWhuykWauJKSGlv5IEXutbI-9UGkJLKLcWktpYSHQPUGW3x5aK2VmccaLY-htmAGceAVt+B2zsGbAALMLM2ABWAAzNw+ANtpEiOGDbMcIjpF8LNuyVwFZ7TeydBwF0YjgKsHZIVWa6AV7d3XgBVY-cQxDzjlnaM48T5TwELXX2wABHAU4J0ExId-x91aAPaOYYbFvzsZXSeX4nEz20cAERjdVLJlev9dibc-gd34qYnxAZLGD0CTbYYtiK6-3eOk3umS-FWMCQfUeITCkzGKSKXxFA+LbF+HtSwTx+BBKqRXCeXcfQAEIzENKaeiTaDIYDOQYIQJExZE4hIno4n8Li7axObiYVCySKCpO3qvHu9TSmNP8bvaxuT8llhqWEzRdcXQAHZgLslYKwdwLgOCoQAHJXxvkaVCVw4kpJyGk7xJTN5ZICbHYA9xBwgHhlEOG1xfkmDSO86+vFgndIceExZs8LYrPiahcFyKGBfIhidDZWyimAr2cCsp2TY4nNRWczM1cIlaKGFbHFLc8UfMYESxJJL4VktqRSsO+zhlHJyXk+lOdGUaOniyl0uT2VrMnA+XwGzVBeLXhkqlBzym0olV0hledBWaqBRY6loLh6nOPuSk1lKzU6u2C+N86BJX2KZZiqJ0jFWvSPLeU855NRqo1bs4V2rRV70ta6pA5zem2tDfa8NLSrDtPUHSg1UqjXBzjRvBNhzGj2GTR0ypsy0U2pDTmiOjq1z8AwFG+ZGLnGzxtmbEAQEA7JFQigAVWby3mMrRa9Q4LfKwv8LC6UdaT6xt7UMvNSa2kdKHZC6FgMoUGAnWWwZIrZ0FvneofVJbDVZgubKq5wBm3sjbdoPUnaiVJH1Eim++hu1Ts3WGvN1jF0jphau5IiKuUMCtW64106t0hjnSmz9q7R0-vvf+wDoSX1atzWBndKb93l0Pe6xtnqzbwBUIhBkGh0BSkQmQPQDJELttel2rIr0GTPp2a+5DNKDiQeXWO-UDJ4M9MY0h-tu9wMLohV+ldjhf1IC4+u4DTH+P5taWh+D5yFnYfrmbOwl7xgduJXdGjwaZMgrFWC4TUHv1icU9Knt+nzUCdQ0J4dJnRNrvTUByzfGDNycLXu8zmblORPrmIjT16nBsXvBedVALs19vcx+4z7Gf3eaPYh01snBODtiwaMd3HJ28eS+51L6Gj4Zh8w2vz8qBGtrUMBLTGQUD-O2XU+NsmYv2bi2ZqTrncvWY87utjGX4vtaS3alLtmvNRqUyVuVZ6BEXsq1RvKtXXo1dgwSurG63Ndea0uvrYnluPoGzlobeWRu9egztv9BKstrc6w69E+WEtYdK1NvDKACNEZIwQ8jlGr3mHm7RvKDJVvSfWzdwzrH0unfMGkSTzmEMHca0d+TdmtsQ-E9Dg91qgfXcTSNgr79xuXJcTbAR6nZvfa07yu6NXAcdcOxtwJJ3TNOfR0VxLcOK0I88wzxz46Yc8Ya+zrrd2xsWd85Nm2IjAs-e04DKnengfhs2yJzLwvM2Dfh4L474PGc8+Z7D-nUWNeI9GzD-HJ7Cd2xUFeFiGzW78qyBKMncusfvoqSPXXfOhUC5BxG1NuPqki4m6em2tyV5K2AIoJAksLSBvhbb1Z2RAhU2yHfPgYeI8YMcFTKA9RhLuBHCBxQl38xMB9xkQ47AICHByPtjPPvy8FDdxhjNR6KRAA)

It's actually a little simpler and easier to explain, so I also included a comment with an overview of how it works.

I'll modify my suggested-changes. (Can you accept them, or can only @ulrichb accept them?)

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+    if (Array.isArray(arg)) arg.push(10); // Should FAIL
+}
+
+function fn9(arg: string | number[] | readonly string[]) {
```
`fn9()` is identical to `fn8()`, how about:

```suggestion
function fn9(arg: string | readonly number[] | string[]) {
```

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+    if (Array.isArray(arg)) arg.manifest; // Should OK
+}
+
+function fn12<T>(arg: T | T[], t: T) {
+    if (Array.isArray(arg)) arg.push(t); // Should OK
+}
```
```suggestion
function fn12a<T>(arg: T | T[], t: T) {
    if (Array.isArray(arg)) arg.push(t); // Should OK
    if (Array.isArray(arg)) arg.indexOf(t); // Should OK
    if (Array.isArray(arg)) arg.indexOf(0); // Should FAIL
}

function fn12b<T>(arg: T | T[] | (T & Number)[], t: T & Number) {
    if (isArrayTest(arg)) arg.push(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}

function fn12c<T, U>(arg: T | T[] | U[], t: T, u: U, tu: T & U) {
    if (isArrayTest(arg)) arg.push(tu); // Should OK
    if (isArrayTest(arg)) arg.indexOf(tu); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}

function fn12d<T>(arg: T | readonly T[], t: T) {
    if (isArrayTest(arg)) arg.push(t); // Should FAIL
    if (isArrayTest(arg)) arg.indexOf(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}
```

laughinghan commented on tests/cases/compiler/isArray.ts:
```diff
@@ -1,9 +1,86 @@
+    if (Array.isArray(arg)) arg.indexOf(t); // OK
+}
+
+function fn14<T>(arg: T | [T], t: T) {
+    if (Array.isArray(arg)) arg.push(t); // Should OK
+}
```
```suggestion
function fn14a<T>(arg: T | [T], t: T) {
    if (isArrayTest(arg)) arg.push(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}

function fn14b<T>(arg: T | [T] | [T & Number], t: T & Number) {
    if (isArrayTest(arg)) arg.push(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}

function fn14c<T, U>(arg: T | [T] | [U], t: T, u: U, tu: T & U) {
    if (isArrayTest(arg)) arg.push(tu); // Should OK
    if (isArrayTest(arg)) arg.indexOf(tu); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}

function fn14d<T>(arg: T | readonly [T], t: T) {
    if (isArrayTest(arg)) arg.push(t); // Should FAIL
    if (isArrayTest(arg)) arg.indexOf(t); // Should OK
    if (isArrayTest(arg)) arg.indexOf(0); // Should FAIL
}
```
